crates/polars-utils/src/pl_str.rs

```diff
@@ -68,70 +69,66 @@ impl Default for PlSmallStr {
     }
 }
 
+/// AsRef, Deref and Borrow impls to &str
+
 impl AsRef<str> for PlSmallStr {
     #[inline(always)]
     fn as_ref(&self) -> &str {
         self.0.as_ref()
     }
 }
 
-impl<T> PartialEq<T> for PlSmallStr
-where
-    T: AsRef<str> + ?Sized,
-{
-    fn eq(&self, other: &T) -> bool {
-        self.as_str() == other.as_ref()
-    }
-}
+impl core::ops::Deref for PlSmallStr {
+    type Target = str;
 
-impl PartialEq<PlSmallStr> for &str {
-    fn eq(&self, other: &PlSmallStr) -> bool {
-        *self == other.as_str()
+    #[inline(always)]
+    fn deref(&self) -> &Self::Target {
+        self.0.as_ref()
     }
 }
 
-impl PartialEq<PlSmallStr> for &&str {
-    fn eq(&self, other: &PlSmallStr) -> bool {
-        **self == other.as_str()
+impl core::borrow::Borrow<str> for PlSmallStr {
+    #[inline(always)]
+    fn borrow(&self) -> &str {
+        self.0.as_ref()
     }
 }
 
-impl PartialEq<PlSmallStr> for String {
-    fn eq(&self, other: &PlSmallStr) -> bool {
-        self.as_str() == other.as_str()
-    }
-}
+/// AsRef impls for other types (TODO)
 
-impl PartialEq<PlSmallStr> for &String {
-    fn eq(&self, other: &PlSmallStr) -> bool {
-        self.as_str() == other.as_str()
+/// From impls
+
+impl From<&str> for PlSmallStr {
+    #[inline(always)]
+    fn from(value: &str) -> Self {
+        Self::from_str(value)
     }
 }
 
-impl core::fmt::Debug for PlSmallStr {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        self.as_str().fmt(f)
+impl From<&&str> for PlSmallStr {
```

Comment:
When do we have double indirection in `&&str`?, I think we should just deref.

Reply:
Yes, please don't add this.

Reply:
We have some functions that take `impl IntoIterator<Item = impl Into<PlSmallStr>>` -

https://github.com/pola-rs/polars/blob/b3172aa411ebcc2eae8adc49b977fe8f7bbdbd34/crates/polars-ops/src/frame/join/mod.rs#L84-L93

and they get passed `&[&str]` - 

https://github.com/pola-rs/polars/blob/b3172aa411ebcc2eae8adc49b977fe8f7bbdbd34/crates/polars/tests/it/core/joins.rs#L508-L515

Whose `IntoIterator` then becomes `&&str`, so we needed the `From<&&str>`.

Reply:
I think we can just pass `["col1", "join_col2"]` instead of `&["col1", "join_col2"]`.

Reply:
That still gives `&&str` for the `IntoIter`

What I can do is to just use `AsRef<str>` instead of `Into<PlSmallStr>` for the publicly exposed API functions. It just means that when we call them internally where we already have `PlSmallStr` it'll allocate a new one

Reply:
@ritchie46 removing this causes a huge amount of breakage, I will do it in a later PR